### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by tshemsedinov**

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    for (const item of arr) {
+      this.addOne(item);
+    }
+  }
+
+  searchByValue(value) {
+    const find = {};
+    if (typeof value === 'string') {
+      for (const hash in this.values) {
+        for (const item in this.values[hash]) {
```
Comment (tshemsedinov) [inline main.js]: - for.of and simple c-for is better then for..in
- you can avaid nested loops here

File: main.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      }
+    } else if (typeof value === 'object') {
+      for (const hash in this.values) {
+        find[hash] = this.values[hash];
+      }
+      for (const hash in this.values) {
+        for (const keys in value) {
+          if (this.values[hash].hasOwnProperty(keys)) {
+            if (this.values[hash][keys] !== value[keys]) {
+              delete find[hash];
```
Comment (tshemsedinov) [inline main.js]: delete is a bad practice

File: main.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    return find;
+  }
+
+  searchByKey(key) {
+    const find = {};
+    for (const hash in this.values) {
+      if (
+        this.values.hasOwnProperty(hash) &&
+                this.values[hash].hasOwnProperty(key)
+      ) {
```
Comment (tshemsedinov) [inline main.js]: if condition should be single-line and simple

File: main.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      this.values[i] = this.values[-(i + 1)];
+      delete this.values[-i];
+    }
+    delete this.values[this.size];
+    delete this.values[-this.size];
+    this.size--;
+  }
+
+  sort(key, reverse) {
+    let temp;
```
Comment (tshemsedinov) [inline main.js]: We don't need `temp` variable here

File: main.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    for (let i = 0; i < Object.keys(this.values).length - 1; i++) {
+      for (let j = 1; j < Object.keys(this.values).length - i; j++) {
+        if (this.values[j][key] > this.values[j + 1][key]) {
+          temp = this.values[j];
+          this.values[j] = this.values[j + 1];
+          this.values[j + 1] = temp;
```
Comment (tshemsedinov) [inline main.js]:
```suggestion
          const temp = this.values[j];
          this.values[j] = this.values[j + 1];
          this.values[j + 1] = temp;
```

File: main.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        }
+      }
+    }
+    if (reverse) {
```
Comment (tshemsedinov) [inline main.js:106]:
```suggestion
    if (!reverse) return;
```

File: main.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+      for (
+        let i = 1, end = Object.keys(this.values).length;
+        i <= end / 2;
+        i++
+      ) {
+        temp = this.values[i];
+        this.values[i] = this.values[end - i + 1];
+        this.values[end - i + 1] = temp;
+      }
```
Comment (tshemsedinov) [inline main.js]: `values` is object, why do you work with it like with array?

File: main.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+class HashTable {
+  constructor() {
+    this.values = {};
+    this.length = 0;
+    this.size = 0;
+  }
+
+  addOne(elem) {
+    const hash = ++this.size;
```
Comment (tshemsedinov) [inline main.js]: We can't just get hash from size
unnamed

File: main.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                this.values[hash].hasOwnProperty(key)
+      ) {
+        find[hash] = this.values[hash][key];
+      }
+    }
+    if (Object.keys(find).length === 0) return null;
+    return find;
+  }
+
+  change(hash, key, value) {
```
Comment (tshemsedinov) [inline main.js]: What is hash here?

File: main.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    if (!this.values.hasOwnProperty(hash)) {
+      this.values[hash] = {};
+    }
+    if (!this.values[hash].hasOwnProperty(key)) {
+      this.length++;
+    }
+    this.values[hash][key] = value;
+  }
+
+  remove(index) {
```
Comment (tshemsedinov) [inline main.js:81]: what is index here?

**Review by tshemsedinov — COMMENTED** [review]
You need to describe data structure first, not clear what do you want to implement

**Comment by SokolyukMisha** [comment]
> You need to describe data structure first, not clear what do you want to implement

have just changed readme file and main.js due to your code review.

File: main.js (hunk 6 above)
Comment (SokolyukMisha) [reply main.js:106]: I think there is a misunderstanding, i describe how it works in readme file

**Pull request merged**